django/core/checks/model_checks.py

```diff
@@ -34,8 +36,25 @@ def check_all_models(app_configs=None, **kwargs):
             errors.extend(model.check(**kwargs))
         for model_index in model._meta.indexes:
             indexes[model_index.name].append(model._meta.label)
+
+        check_constraints_seen = {}
         for model_constraint in model._meta.constraints:
             constraints[model_constraint.name].append(model._meta.label)
+            if isinstance(model_constraint, CheckConstraint):
+                if model_constraint.check in check_constraints_seen:
```

Comment:
I'm not sure is this really helpful and needed :thinking:  because it'll catch only checks with the same order of conditions, so only copy&paste issues. For example, it will not raise a warning for:
```python
models.CheckConstraint(check=models.Q(models.Q(id__gt=0) | models.Q(id=0))
models.CheckConstraint(check=models.Q(models.Q(id=0) | models.Q(id__gt=0))
```
(that's of course only a simple example). In the same time I would not like to add any complicated logic here.

Reply:
Thanks for the feedback Mariusz. Just so I'm clear, are you suggesting this PR be closed or are you suggesting this PR be modified to catch these kinds of duplicates as well?

Reply:
Honestly, I'm not in favor of another system check, but I'll try to confirm this with Carlton tomorrow (who accepted this ticket).

Reply:
> I'm not in favor of another system check

Could you explain the reason why? (Assuming we can solve the ordering concern already mentioned.)

Reply:
It is not only about ordering. Condition can be more complicated, e.g.
```python
check=Q(Q(field_1='a') & Q(Q(field_2='b') | Q(field_3='c')))
check=Q(Q(Q(field_3='c') | Q(field_2='b')) & Q(field_1='a')) 
```
Moreover they can be logically equivalent, e.g.
```python
check=~Q(Q(field_1='a') | Q(field_2='b'))
check=Q(~Q(field_1='a') & ~Q(field_2='b'))
```
etc.

Users can assume that Django protects them against duplicate constraints. Furthermore I'm afraid that someone somewhere in the universe :alien: uses a subclass of `CheckConstraint` where the same conditions make sense (we had regressions in many checks).

Reply:
Check constraints can add significant overhead to insert/update queries. Adding duplicates (whether identical or in a different form) compounds that and is probably worth highlighting.

I wouldn't have thought that dealing with ordering and logical equivalence was a deal breaker. Yes, it requires some thought, but a well-defined `simplify_q()` function that can normalize a `Q()` object would make sense.

I also don't see this being limited to check constraints - partial indexes can also make use of `Q()` objects for the `condition`. Duplicate indexes also have an impact on insert/update as well as wasting valuable disk space.

Reply:
>  Furthermore I'm afraid that someone somewhere in the universe uses a subclass of `CheckConstraint` where the same conditions make sense (we had regressions in many checks).

I agree. I intentionally chose using a `Warning` and not an `Error` for this reason. It isn't technically invalid, but normally points to a programming oddity worth double checking.

Reply:
> ... Carlton tomorrow (who accepted this ticket).

OK, so _Accepted_ just means, _Let's have a look at it_, not _It's definitely going in_ — I know we all know that, but totally happy for you to reject patches @felixxm! 🙂

Then... 

The trade-off here: Are we going to help more folks that we annoy? I don't know what the answer is there. 
(It seemed a reasonable enough suggestion at the time.) 

* We **have** had regressions caused by system checks, so I do understand the caution. 
* Folks can always silence it? 
* Or add these kind of things to their own project? 

>  … a well-defined simplify_q() function that can normalize a Q() object would make sense.

Is that something that might appear some time soon, or is it just a thought? 🙂

Reply:
> Is that something that might appear some time soon, or is it just a thought? 🙂

At the moment just a thought… 🙂

Reply:
Interesting topic, I recently had the pleasure to debug a very large query and at the end the problem was caused by a duplicate condition on using a subquery. As much as I would have loved to have had this check tell me about the problem, I don't believe it's possible to build it. Mainly because it's really hard to build a good `simplify_q`method. Especially since two queries may yield the same result, but one uses and index the other does not.
Furthermore, it will be hard outperform the query optimizer that comes with each database. I am also not sure that this should be our goal. I do agree with @felixxm point, that this will provide people with a false sense of security. There will probably also be a long line of patches for edge cases, once the feature is released.

Not saying you shouldn't try, just saying it will be really really hard.

Reply:
@codingjoe I'm not sure we're taking about the same thing. This is about avoiding duplicate check constraints being defined on a model.

Reply:
Thanks @pope1ni, i really misunderstood the subject. I removed my comment to avoid further confusion.